### ISSUE TITLE
Test timeouts with Dart 2.0.0-dev.64.2

### ISSUE THREAD

**Issue by srawlins** [dart-lang/vm_service_client#32]
Not sure when these started, but `pub run test` fails.

* `dart test/bound_field_test.dart` times out.
* `dart test/function_test.dart` times out.
* `dart test/instance_test.dart` times out.
* `dart test/source_report_test.dart` times out.
* ...

**Comment by DanTup** [comment]
Seems like they're timing out on this:

```dart
await isolate.waitUntilPaused()
```

As far as I can see, no response was received to `getIsolate`:

```
{"jsonrpc":"2.0","method":"getVM","id":0,"params":{}}
{"jsonrpc":"2.0", "result":{"type":"VM","name":"vm","architectureBits":64,"targetCPU":"x64","hostCPU":"Intel(R) Core(TM) i7-7700HQ CPU @ 2.80GHz","version":"2.0.0-dev.65.0 (Tue Jun 26 14:17:21 2018 +0200) on \"macos_x64\"","_profilerMode":"Dart","_nativeZoneMemoryUsage":0,"pid":23345,"startTime":1531303603168,"_embedder":"Dart VM","_maxRSS":176091136,"_currentRSS":171266048,"isolates":[{"type":"@Isolate","fixedId":true,"id":"isolates\/638066918","name":"%20%20%20%20%20new%20ReceivePort();%20%20%20%0A%0A%20%20%20%20%20%20%20%20var%20foo%20=%20new%20Foo();%0A%20%20%20%20%20%20debugger();%0A%20%20%20%20%0A%7D%0A:main()","number":"638066918"}]},"id":0}
{"jsonrpc":"2.0","method":"streamListen","id":1,"params":{"streamId":"Debug"}}
{"jsonrpc":"2.0","method":"getIsolate","id":2,"params":{"isolateId":"isolates/638066918"}}
{"jsonrpc":"2.0","id":1,"result":{"type":"Success"}}
// HANG
^C^C
```

**Comment by DanTup** [comment]
Looks like these tests are running code like this:

> /Users/dantup/Dev/dart-sdk/v2/bin/dart
[--pause-isolates-on-exit, --enable-vm-service=0, data:application/dart;charset=utf-8,import%20%22dart:async%22;%20import%20%22dart:developer%22;%20import%20%22dart:io%22;%20import%20%22dart:isolate%22;%20import%20%22dart:typed_data%22;%20import%20%22dart:mirrors%22;%0A%0A%20%20%20%20%20%20class%20Foo%20%7B%0A%20%20%20%20%20%20%20%20final%20int%20val%20=%2010;%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%0A%0Amain()%20async%20%7B%0A%20%20%20%20%20%20/*%20Wait%20for%20a%20line%20so%20the%20test%20doesn't%20print%20something%20that%20could%20be%20emitted%20%20%20%20%20%20*%20before%20%22Observatory%20listening%20on%22.%20*/%20%20%20%20%20stdin.readLineSync();%20%20%20%20%20%20/*%20Don't%20let%20the%20isolate%20close%20on%20its%20own.%20*/%20%20%20%20%20new%20ReceivePort();%20%20%20%0A%0A%20%20%20%20%20%20%20%20var%20foo%20=%20new%20Foo();%0A%20%20%20%20%20%20debugger();%0A%20%20%20%20%0A%7D%0A]

And then we get this output:

> Observatory listening on http://127.0.0.1:60674/
> data:application/dart;charset=utf-8,import%20%22dart:async%22;%20import%20%22dart:developer%22;%20import%20%22dart:io%22;%20import%20%22dart:isolate%22;%20import%20%22dart:typed_data%22;%20import%20%22dart:mirrors%22;%0A%0A%20%20%20%20%20%20class%20Foo%20%7B%0A%20%20%20%20%20%20%20%20final%20int%20val%20=%2010;%0A%20%20%20%20%20%20%7D%0A%20%20%20%20%0A%0Amain()%20async%20%7B%0A%20%20%20%20%20%20/*%20Wait%20for%20a%20line%20so%20the%20test%20doesn't%20print%20something%20that%20could%20be%20emitted%20%20%20%20%20%20*%20before%20%22Observatory%20listening%20on%22.%20*/%20%20%20%20%20stdin.readLineSync();%20%20%20%20%20%20/*%20Don't%20let%20the%20isolate%20close%20on%20its%20own.%20*/.packages:1: Error: Unable to read '.packages' file:
>   FormatException: Not a valid package name (at character 1)
> import "dart
> ^
> .

The VM terminates, and therefore there's never a response to the requests sent over the websocket (I guess the web socket being terminated immediately is not handled well by the tests).

@devoncarew @jacob314 Any ideas about this? Is the behaviour of passing `data:application/dart` in place of a script name documented (I've looked, but can't find anything). I'm not sure if this is an accidental break in the VM or something deliberate that just needs a fix to the tests?

**Comment by DanTup** [comment]
With a simpler script:

```text
dantup-macbookpro:vm_service_client dantup$ dart "data:application/dart;charset=utf-8,main%28%29%7Bprint%28%22test%22%29%3B%7D"
data:application/.packages:1: Error: Unable to read '.packages' file:
  FormatException: Expecting '='
application/.packages.
```

In v1 this would work:

```text
dantup-macbookpro:vm_service_client dantup$ ~/Dev/dart-sdk/v1/bin/dart "data:application/dart;charset=utf-8,main%28%29%7Bprint%28%22test%22%29%3B%7D"
test
```

**Comment by devoncarew** [comment]
@kevmoo @a-siva @mraleph - is the fact that we can't run from `data:` urls in dart 2 expected, or an unintentional change?

**Comment by mraleph** [comment]
To the best of my knowledge we are expected to be able to run from `data:` URLs - if this got broken - then it is a bug.

**Comment by DanTup** [comment]
Ok thanks for confirming, I'll open an SDK issue.

**Comment by DanTup** [comment]
I've opened https://github.com/dart-lang/sdk/issues/34639 (and tagged with `area-vm` - hope that's right!).

**Comment by mraleph** [comment]
I think the fact that you could run `dart "data:..."` was a bit accidental though. (`dart --help` does not specify whether it expects file or a uri) so the fact that you _could_ use `data:` uri to begin with was accidental.

**Comment by DanTup** [comment]
Ok; is there someone well-placed to make a decision about whether it should be restored or considered not supported? If it's not supported, we can look at changing these tests to not use it (which may be as simple as writing to a temp folder, but I'm not familiar enough to bet on it) but if it's likely to come back it might not be worth it.

**Comment by DanTup** [comment]
In the absence of any info on whether it's likely the VM should/will support this; I may take a stab at making the tests work without using the data: uri next week. I suspect writing to a temp folder will be easiest, but if anyone has any suggestions/preferences for this, please let me know (cc @nex3)

**Comment by mraleph** [comment]
@DanTup yeah, I think this is extremely obscure feature of the Dart VM which was an accident - so I think it is safe to say that we are not going to fix this.

**Comment by devoncarew** [comment]
> so the fact that you could use data: uri to begin with was accidental

In lieu of additional guidance from @nex3 here, I think we should convert the tests over to running from file: uris, and just store them in a temporary directory or similar.

For context, we need changes to this this package in order to unblock some tests for flutter_tools. We'd like to get the tests here running again in order to have confidence in our PRs here.

**Comment by DanTup** [comment]
Changing this over to write to a file and updating some assumptions about the scheme being `data` fixes some issues, however there are lots of other failures that I don't entirely understand. I'm not sure if they're broken by this change, or something else in all the VM changes. For example the tests checking frames/call stacks all get unexpected results (for ex. expecting `<main_async_body>` or `main.<fn>` but getting `async_op`).

<img width="727" alt="screen shot 2018-10-10 at 11 22 11 am" src="https://user-images.githubusercontent.com/1078012/46730762-c6eb7500-cc80-11e8-8a6c-4057717063a3.png">

<img width="855" alt="screen shot 2018-10-10 at 11 26 03 am" src="https://user-images.githubusercontent.com/1078012/46730770-ceab1980-cc80-11e8-8037-c3d9679f9d17.png">

Here's a stack with the locations added:

<img width="867" alt="screen shot 2018-10-10 at 11 36 16 am" src="https://user-images.githubusercontent.com/1078012/46730789-d965ae80-cc80-11e8-9775-531f2dcce073.png">

Where did the `start` and `async_op` frames come from? And why don't we have `main.<fn>` anywhere? Are these valid VM changes and these tests are just outdated?

**Comment by mraleph** [comment]
I think `async_op` is the new name of `main_async_body`. `start` is a helper function which is part of async await implementation. I think there is a logic to hide some of these frames from the debugger @aam was recently fixing it up - so he might have a better understanding of what is expected from this test.

**Comment by DanTup** [comment]
@mraleph Ah, that could make some sense. The tests were checking:

1. `frame.toString() == #0 in <main_async_body>`
2. `frame.member == 'main.<fn>'`

And it looks like now we have:

1. `frame.toString() == #0 in async_op`
2. `frame.member == 'main.async_op'`

Changing those things makes some more passes, so if @aam thinks that sounds sane, I'll do that.

What's slightly strange though, is in the stack trace from the test failure, it has `main.<fn>`!

<img width="473" alt="screen shot 2018-10-10 at 12 50 08 pm" src="https://user-images.githubusercontent.com/1078012/46734457-28184600-cc8b-11e8-8904-e1f96755dc02.png">

**Comment by aam** [comment]
> For example the tests checking frames/call stacks all get unexpected results (for ex. expecting <main_async_body> or main.<fn> but getting async_op).

I think the reason for this is change in stack trace is that  in dart2 async functions are invoked and executed synchronously(so called "sync-async" feature), going into proper async-machinery on first await only.
